dtfs-central-api/src/v1/controllers/utilisation-report-service/get-utilisation-report-reconciliation-details-by-id.controller/helpers/get-keying-sheet-for-report-id.test.ts

```diff
@@ -0,0 +1,405 @@
+          feeRecord: {
+            report: { id: reportId },
+            status: In<FeeRecordStatus>(['READY_TO_KEY', 'RECONCILED']),
+          },
+        },
+        relations: {
+          feeRecord: true,
+          payment: true,
+        },
+      })
+      .mockResolvedValue(joinTableEntities);
+
+    // Act
+    const keyingSheet = await getKeyingSheetForReportId(reportId, [feeRecord]);
+
+    // Assert
+    expect(keyingSheet).toHaveLength(1);
+    expect(keyingSheet[0].feePayments).toHaveLength(3);
+    expect(keyingSheet[0].feePayments[0].amount).toBe(1000);
+    expect(keyingSheet[0].feePayments[1].amount).toBe(2000);
+    expect(keyingSheet[0].feePayments[2].amount).toBe(3000);
+  });
+
+  it('sets the keying sheet row fee payment currency and date received to the corresponding payment value', async () => {
+    // Arrange
+    const firstPayment = PaymentEntityMockBuilder.forCurrency('GBP').withId(11).withDateReceived(new Date('2021')).build();
+    const secondPayment = PaymentEntityMockBuilder.forCurrency('USD').withDateReceived(new Date('2022')).withId(22).build();
+    const thirdPayment = PaymentEntityMockBuilder.forCurrency('EUR').withDateReceived(new Date('2023')).withId(33).build();
+
+    const feeRecord = FeeRecordEntityMockBuilder.forReport(aUtilisationReport()).withStatus('READY_TO_KEY').withId(12).build();
+
+    const joinTableEntities = [
+      aFeeRecordPaymentJoinTableEntityWith({ feeRecord, payment: firstPayment, paymentAmountUsedForFeeRecord: 1000 }),
+      aFeeRecordPaymentJoinTableEntityWith({ feeRecord, payment: secondPayment, paymentAmountUsedForFeeRecord: 2000 }),
+      aFeeRecordPaymentJoinTableEntityWith({ feeRecord, payment: thirdPayment, paymentAmountUsedForFeeRecord: 3000 }),
+    ];
+
+    const reportId = 1;
+    when(mockFind)
+      .calledWith(FeeRecordPaymentJoinTableEntity, {
+        where: {
+          feeRecord: {
+            report: { id: reportId },
+            status: In<FeeRecordStatus>(['READY_TO_KEY', 'RECONCILED']),
+          },
+        },
+        relations: {
+          feeRecord: true,
+          payment: true,
+        },
+      })
+      .mockResolvedValue(joinTableEntities);
+
+    // Act
+    const keyingSheet = await getKeyingSheetForReportId(reportId, [feeRecord]);
+
+    // Assert
+    expect(keyingSheet).toHaveLength(1);
+    expect(keyingSheet[0].feePayments).toHaveLength(3);
+    expect(keyingSheet[0].feePayments[0].currency).toBe('GBP');
+    expect(keyingSheet[0].feePayments[0].dateReceived).toEqual(new Date('2021'));
+    expect(keyingSheet[0].feePayments[1].currency).toBe('USD');
+    expect(keyingSheet[0].feePayments[1].dateReceived).toEqual(new Date('2022'));
+    expect(keyingSheet[0].feePayments[2].currency).toBe('EUR');
+    expect(keyingSheet[0].feePayments[2].dateReceived).toEqual(new Date('2023'));
+  });
+
+  it('sets the keying sheet row fee payments only to the payment with a non-null paymentAmountUsedForFeeRecord column', async () => {
```

Comment:
Could we add a test for a fee record, which has payments, but all the fee payments are null, that checks it gets set to a zero payment, no date, keying sheet row? The logic looks correct and should get handled similarly to the no payments case below, but would be goof to have test coverage.